### PR TITLE
#50102 Task 1: Move apimachinery/pkg/watch.Until into client-go/tools/watch.UntilWithoutRetry

### PR DESCRIPTION
**Pull request by tnozicka**
**What this PR does / why we need it**:
This is a split off from https://github.com/kubernetes/kubernetes/pull/50102 to go in smaller pieces.

Moves `apimachinery/pkg/watch.Until` into `client-go/tools/watch.UntilWithoutRetry` and adds context so it is cancelable.

**Release note**:
```release-note
NONE
```

**Dev release note**:
```dev-release-note
`apimachinery/pkg/watch.Until` has been moved to `client-go/tools/watch.UntilWithoutRetry`.
While switching please consider using the new `client-go/tools/watch.UntilWithSync` or `client-go/tools/watch.Until`.
```

/cc @smarterclayton @kubernetes/sig-api-machinery-pr-reviews 
/milestone v1.12
/priority important-soon
/kind bug
(bug after the main PR which is this split from)

### CODE REVIEW

**Comment by sttts** [comment]
Didn't we discuss this before? Why this move?

/hold

**Comment by sttts** [comment]
Also many people use this helper I would expect. This move breaks them.

**Comment by tnozicka** [comment]
> Didn't we discuss this before? 

I recall discussing where to put it, thought client-go was what you recommended unless I got you wrong.

> Why this move?

Renaming was the consensus between reviewers on https://github.com/kubernetes/kubernetes/pull/50102. 

The move is necessary because of import restrictions and because the other Until* functions need to include informers.

a) moving the package breaks them anyways
b) this function is broken and going away (internal only) to be replaced with one that can retry failed watches in next PR

I don't want to break them twice by moving the function for v1.12 and renaming it in 1.13 where the rest of usages will be replaced and also having to rename the Retry function back to Until. That's why also the rename now since we are already breaking them with package move.

**Comment by sttts** [comment]
/cancel hold

Missed this reasoning in the PR description.

Please provide a `dev-release-note`.

**Comment by tnozicka** [comment]
> Please provide a `dev-release-note`.

@sttts would you mind giving me a pointer on how that should look like? Is that a regular release note with some prefix/convention? thx

**Comment by tnozicka** [comment]
rebased again...

AFAIK @smarterclayton and @wojtek-t are out. 
/assign @liggitt 
seems like you are the only one left from the main reviewers of the PR this is split from, possibly @sttts 

I'd like to ship this as there are and will be more PRs depending on this instead of keeping to rebase them all every day.

/cc @kubernetes/sig-api-machinery-pr-reviews

**Comment by sttts** [comment]
> would you mind giving me a pointer on how that should look like?

Yes, just the `dev-release-note` tag. Makes it much easier to compile client-go release notes later.

**Comment by tnozicka** [comment]
@sttts updated the PR description to contain `dev-release-note`

File: staging/src/k8s.io/client-go/tools/watch/until.go [hunk 1]
```diff
@@ -17,38 +17,39 @@ limitations under the License.
...
-var ErrWatchClosed = errors.New("watch closed before Until timeout")
+// ErrWatchClosed is returned when the watch channel is closed before timeout in UntilWithoutRetry.
+var ErrWatchClosed = errors.New("watch closed before UntilWithoutRetry timeout")
 
-// Until reads items from the watch until each provided condition succeeds, and then returns the last watch
+// UntilWithoutRetry reads items from the watch until each provided condition succeeds, and then returns the last watch
 // encountered. The first condition that returns an error terminates the watch (and the event is also returned).
 // If no event has been received, the returned event will be nil.
 // Conditions are satisfied sequentially so as to provide a useful primitive for higher level composition.
 // A zero timeout means to wait forever.
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/watch/until.go]: fix doc to match context arg ("waits until context deadline or until context is canceled", etc)

**Comment by liggitt** [comment]
/lgtm

can follow up with the doc tweak in the next PR in the series

**Comment by tnozicka** [comment]
fixed the nit in the comment, thx @liggitt

**Comment by tnozicka** [comment]
@lavalamp ptal, needs approval for `staging/`

**Pull request merged**